packages/entity-server/src/entity/middleware.ts

```diff
@@ -0,0 +1,101 @@
+/**
+ * Tupaia
+ * Copyright (c) 2017 - 2021 Beyond Essential Systems Pty Ltd
+ */
+
+import { NextFunction, Response } from 'express';
+import { AccessPolicy } from '@tupaia/access-policy';
+import { PermissionsError } from '@tupaia/utils';
+import { EntityModel, EntityFields, FetchEntityRequest, FetchEntityResponse } from '../types';
+
+export const attachEntityToRequest = async (
+  req: FetchEntityRequest,
+  res: Response,
+  next: NextFunction,
+) => {
+  try {
+    const { entityCode, projectCode } = req.params;
+    const entity = await req.models.entity.findOne({
+      code: entityCode,
+    });
+
+    if (!entity || !(await userHasAccessToEntity(entity, projectCode, req.accessPolicy))) {
+      throw new PermissionsError(`No access to requested entity: ${entityCode}`);
+    }
+
+    req.context.entity = entity;
+
+    next();
+  } catch (error) {
+    next(error);
+  }
+};
+
+export const attachFormatterToResponse = async (
+  req: FetchEntityRequest,
+  res: FetchEntityResponse,
+  next: NextFunction,
+) => {
+  try {
+    res.context.formatEntityForResponse = mapEntitiesToFields(
+      extractFieldsFromQuery(req.query.fields),
+    );
+    next();
+  } catch (error) {
+    next(error);
+  }
+};
+
+const userHasAccessToEntity = async (
+  entity: EntityModel,
+  projectCode: string,
+  accessPolicy: AccessPolicy,
+) => {
+  // Assume user always has access to all world items.
+  if (entity.code === 'World') {
+    return true;
+  }
+  // Timor-Leste is temporarily turned off
+  if (entity.country_code === 'TL') {
+    return false;
+  }
+
+  // project access rights are determined by their children
+  if (entity.isProject()) {
+    const projectChildren = await entity.getChildren(projectCode);
+
+    return accessPolicy.allowsSome(projectChildren.map(c => c.country_code));
+  }
+
+  return accessPolicy.allows(entity.country_code);
+};
+
+const mapEntitiesToFields = (fields: (keyof EntityFields)[]) => (model: EntityModel) => {
+  const mappedEntity: { -readonly [field in keyof EntityFields]?: EntityFields[field] } = {};
+  fields.forEach(field => {
+    mappedEntity[field] = model[field];
+  });
+  return mappedEntity;
+};
+
+const validFields: (keyof EntityFields)[] = ['code', 'country_code'];
```

Comment:
This is rather handly as TS will throw an error here if we try to add a string that isn't a key of `EntityFields` 👍 

Technically there's another syntax that will produce an array of strings of all the keys of `EntityFields` however I opted not to use it for a couple reasons:
- It doesn't read as clearly
- We may wish to only support a subset of the `EntityFields` (I suspect this is likely)